src/main/java/org/opensearch/timeseries/feature/SearchFeatureDao.java

```diff
@@ -210,6 +215,48 @@ public void getLatestDataTime(
         }
     }
 
+    public void getDateRangeOfSourceData(
+        Config config,
+        User user,
+        Map<String, Object> topEntity,
+        ActionListener<Pair<Long, Long>> internalListener
+    ) {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+            .aggregation(AggregationBuilders.min(CommonName.AGG_NAME_MIN_TIME).field(config.getTimeField()))
+            .aggregation(AggregationBuilders.max(CommonName.AGG_NAME_MAX_TIME).field(config.getTimeField()))
+            .size(0);
+        BoolQueryBuilder internalFilterQuery = QueryBuilders.boolQuery().must(config.getFilterQuery());
+        topEntity.entrySet().forEach(entity -> internalFilterQuery.must(new TermQueryBuilder(entity.getKey(), entity.getValue())));
+
+        searchSourceBuilder.query(internalFilterQuery);
+
+        SearchRequest request = new SearchRequest().indices(config.getIndices().toArray(new String[0])).source(searchSourceBuilder);
+        final ActionListener<SearchResponse> searchResponseListener = ActionListener.wrap(r -> {
+            InternalMin minAgg = r.getAggregations().get(CommonName.AGG_NAME_MIN_TIME);
+            InternalMax maxAgg = r.getAggregations().get(CommonName.AGG_NAME_MAX_TIME);
+            double minValue = minAgg.getValue();
+            double maxValue = maxAgg.getValue();
+            // If time field not exist or there is no value, will return infinity value
+            if (minValue == Double.POSITIVE_INFINITY) {
+                internalListener.onFailure(new ResourceNotFoundException(config.getId(), "There is no data in the time field"));
+                return;
+            }
+            internalListener.onResponse(Pair.of((long) minValue, (long) maxValue));
+        }, e -> { internalListener.onFailure(e); });
+
+        // inject user role while searching.
+        clientUtil
+            .<SearchRequest, SearchResponse>asyncRequestWithInjectedSecurity(
+                request,
+                client::search,
+                // user is the one who triggered the caller of this function
+                user,
+                client,
+                AnalysisType.AD,
```

Comment:
is this method just used by AD?

Reply:
Good catch. Changed to

```
clientUtil
            .<SearchRequest, SearchResponse>asyncRequestWithInjectedSecurity(
                request,
                client::search,
                // user is the one who triggered the caller of this function
                user,
                client,
                config instanceof AnomalyDetector ? AnalysisType.AD : AnalysisType.FORECAST,
                searchResponseListener
            );
```

Reply:
double checked other places. Seems fine.